In [1]:
## Import pandas
import pandas as pd

In [2]:
## Import dataset 
data = pd.read_csv("data_gov.csv")
data.head() # Show top five row

Category of States Non Special Category States  \
0  Non Special Category States              Andhra Pradesh   
1  Non Special Category States                       Assam   
2  Non Special Category States                       Bihar   
3  Non Special Category States                Chhattisgarh   
4  Non Special Category States                       Delhi   

   Scheduled Castes - 2004-05  Scheduled Castes - 2007-08  \
0                        17.1                        12.8   
1                         9.4                         9.5   
2                        10.6                        10.0   
3                        12.5                         4.3   
4                         NaN                         NaN   

   Scheduled Tribes - 2004-05  Scheduled Tribes - 2007-08  \
0                        12.0                         8.2   
1                         4.9                         7.7   
2                        11.2                         7.7   
3                         4.2                         5.8   
4                         NaN                         NaN   

   Other Backward Classes - 2004-05  Other Backward Classes - 2007-08  \
0                              10.1                               9.7   
1                               4.3                               6.9   
2                               5.2                               7.0   
3                              10.2                               8.2   
4                               1.5                               4.8   

   Others - 2004-05  Others - 2007-08  All Social Groups - 2004-05  \
0               7.4               5.6                         11.2   
1               7.3               6.4                          6.4   
2               5.9               5.3                          6.8   
3               8.1               9.0                          8.2   
4               2.6               2.9                          1.9   

   All Social Groups - 2007-08  
0                          9.5  
1                          7.0  
2                          7.4  
3                          6.7  
4                          3.8

In [3]:
# Show bottom five rows
data.tail()

Category of States Non Special Category States  Scheduled Castes - 2004-05  \
31  Union Territories        Dadra & Nagar Haveli                         NaN   
32  Union Territories                 Daman & Diu                         NaN   
33  Union Territories                 Lakshadweep                         NaN   
34  Union Territories                  Puducherry                        36.0   
35          All India                   All India                        12.0   

    Scheduled Castes - 2007-08  Scheduled Tribes - 2004-05  \
31                         NaN                         7.1   
32                         NaN                         NaN   
33                         NaN                        25.1   
34                        28.7                         NaN   
35                        11.9                         6.5   

    Scheduled Tribes - 2007-08  Other Backward Classes - 2004-05  \
31                         8.4                               NaN   
32                         NaN                               NaN   
33                        17.0                               NaN   
34                         NaN                              22.2   
35                         7.5                               7.7   

    Other Backward Classes - 2007-08  Others - 2004-05  Others - 2007-08  \
31                               NaN               NaN               NaN   
32                               NaN               NaN               NaN   
33                               NaN               NaN               NaN   
34                              22.8               NaN               NaN   
35                               7.9               6.6               6.4   

    All Social Groups - 2004-05  All Social Groups - 2007-08  
31                          6.0                          6.2  
32                          1.6                          1.7  
33                         22.5                         16.7  
34                         27.4                         25.2  
35                          8.2                          8.4

In [4]:
## shape of dataset
data.shape

(36, 12)

there are 36 no. of rows and 12 no. of columns in our dataset.

In [5]:
## size of data set
data.size

432

there are 432 no. of records in our dataset.

In [6]:
### Checking the data type of columns
data.dtypes

Category of States                   object
Non Special Category States          object
Scheduled Castes - 2004-05          float64
Scheduled Castes - 2007-08          float64
Scheduled Tribes - 2004-05          float64
Scheduled Tribes - 2007-08          float64
Other Backward Classes - 2004-05    float64
Other Backward Classes - 2007-08    float64
Others - 2004-05                    float64
Others - 2007-08                    float64
All Social Groups - 2004-05         float64
All Social Groups - 2007-08         float64
dtype: object

In [7]:
#### Fetch Information from dataset
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36 entries, 0 to 35
Data columns (total 12 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Category of States                36 non-null     object 
 1   Non Special Category States       36 non-null     object 
 2   Scheduled Castes - 2004-05        26 non-null     float64
 3   Scheduled Castes - 2007-08        26 non-null     float64
 4   Scheduled Tribes - 2004-05        27 non-null     float64
 5   Scheduled Tribes - 2007-08        27 non-null     float64
 6   Other Backward Classes - 2004-05  27 non-null     float64
 7   Other Backward Classes - 2007-08  27 non-null     float64
 8   Others - 2004-05                  29 non-null     float64
 9   Others - 2007-08                  29 non-null     float64
 10  All Social Groups - 2004-05       36 non-null     float64
 11  All Social Groups - 2007-08       36 non-null     float64
dtypes: float64

In [8]:
## columns in dataset
data.columns

Index(['Category of States', 'Non Special Category States',
       'Scheduled Castes - 2004-05', 'Scheduled Castes - 2007-08',
       'Scheduled Tribes - 2004-05', 'Scheduled Tribes - 2007-08',
       'Other Backward Classes - 2004-05', 'Other Backward Classes - 2007-08',
       'Others - 2004-05', 'Others - 2007-08', 'All Social Groups - 2004-05',
       'All Social Groups - 2007-08'],
      dtype='object')

In [9]:
## Fetching data for a specific column
data["Category of States"]

0     Non Special Category States
1     Non Special Category States
2     Non Special Category States
3     Non Special Category States
4     Non Special Category States
5     Non Special Category States
6     Non Special Category States
7     Non Special Category States
8     Non Special Category States
9     Non Special Category States
10    Non Special Category States
11    Non Special Category States
12    Non Special Category States
13    Non Special Category States
14    Non Special Category States
15    Non Special Category States
16    Non Special Category States
17    Non Special Category States
18    Non Special Category States
19        Special Category States
20        Special Category States
21        Special Category States
22        Special Category States
23        Special Category States
24        Special Category States
25        Special Category States
26        Special Category States
27        Special Category States
28        Special Category States
29            

In [10]:
### collecting unique value for a specific olumn
data["Category of States"].unique()

array(['Non Special Category States', 'Special Category States',
       'Union Territories', 'All India'], dtype=object)

In [11]:
### fetching 'Scheduled Castes - 2004-05', 'Scheduled Castes - 2007-08', 'Scheduled Tribes - 2004-05', 'Scheduled Tribes - 2007-08' from dataset
data[['Scheduled Castes - 2004-05', 'Scheduled Castes - 2007-08','Scheduled Tribes - 2004-05', 'Scheduled Tribes - 2007-08']]

Scheduled Castes - 2004-05  Scheduled Castes - 2007-08  \
0                         17.1                        12.8   
1                          9.4                         9.5   
2                         10.6                        10.0   
3                         12.5                         4.3   
4                          NaN                         NaN   
5                         19.2                         0.3   
6                          5.4                         3.6   
7                         10.4                        12.2   
8                          5.9                        16.7   
9                          9.8                         8.5   
10                        35.0                        27.2   
11                        12.8                        10.2   
12                        17.6                        10.5   
13                        11.1                         9.8   
14                        14.1                        13.1   
15                         9.1                         8.7   
16                        22.9                        25.1   
17                         5.3                         8.4   
18                        12.5                        13.5   
19                         NaN                         NaN   
20                         8.4                         8.8   
21                         2.9                         5.2   
22                         NaN                         NaN   
23                         NaN                         NaN   
24                         NaN                         NaN   
25                         NaN                         NaN   
26                         8.0                        10.3   
27                        15.3                        19.7   
28                         6.0                        11.0   
29                         NaN                         NaN   
30                        13.4                         4.2   
31                         NaN                         NaN   
32                         NaN                         NaN   
33                         NaN                         NaN   
34                        36.0                        28.7   
35                        12.0                        11.9   

    Scheduled Tribes - 2004-05  Scheduled Tribes - 2007-08  
0                         12.0                         8.2  
1                          4.9                         7.7  
2                         11.2                         7.7  
3                          4.2                         5.8  
4                          NaN                         NaN  
5                          NaN                         NaN  
6                          4.4                         3.5  
7                          NaN                         NaN  
8                          6.9                        10.5  
9                          4.7                         4.9  
10                        36.5                        29.0  
11                         3.3                         9.0  
12                        11.2                        12.2  
13                         6.3                         6.4  
14                         NaN                         NaN  
15                         2.3                         4.5  
16                        26.4                        26.9  
17                         8.4                        11.1  
18                        14.4                        14.3  
19                         1.6                         3.4  
20                         3.2                         4.1  
21                         3.3                        10.1  
22                         0.2                         3.4  
23                         0.4                         1.4  
24                         0.6                         0.6  
25                         3.3                         7.1  
26                         2.3                         3.1  
2

#### Statistical Analysis of dataset using pandas

In [12]:
## mean of "Scheduled Castes - 2004-05" column
data["Scheduled Castes - 2004-05"].mean()

13.180769230769231

In [13]:
## Variance of Scheduled Castes - 2004-05 column
data["Scheduled Castes - 2004-05"].var()

64.55761538461539

In [14]:
## maximum value of Scheduled Castes - 2004-05 column
data["Scheduled Castes - 2004-05"].max()

36.0

In [15]:
## minimum value of Scheduled Castes - 2004-05 column
data["Scheduled Castes - 2004-05"].min()

2.9

***If u want to find Statistical analysis for a complete dataset***

The statistics that are generated by the describe() method:

1) count tells us the number of NoN-empty rows in a feature.

2) mean tells us the mean value of that feature.

3) std tells us the Standard Deviation Value of that feature.

4) min tells us the minimum value of that feature.

5) 25%, 50%, and 75% are the percentile/quartile of each features. This quartile information helps us to detect Outliers.

6) max tells us the maximum value of that feature.

In [16]:
data.describe()

Scheduled Castes - 2004-05  Scheduled Castes - 2007-08  \
count                   26.000000                   26.000000   
mean                    13.180769                   11.700000   
std                      8.034775                    6.955171   
min                      2.900000                    0.300000   
25%                      8.575000                    8.550000   
50%                     11.550000                   10.250000   
75%                     15.000000                   13.025000   
max                     36.000000                   28.700000   

       Scheduled Tribes - 2004-05  Scheduled Tribes - 2007-08  \
count                   27.000000                   27.000000   
mean                     8.411111                    9.111111   
std                      8.793544                    6.959572   
min                      0.200000                    0.600000   
25%                      3.250000                    4.300000   
50%                      4.900000                    7.700000   
75%                     11.200000                   10.800000   
max                     36.500000                   29.000000   

       Other Backward Classes - 2004-05  Other Backward Classes - 2007-08  \
count                         27.000000                         27.000000   
mean                           9.207407                          8.481481   
std                            7.688049                          4.774696   
min                            1.500000                          3.400000   
25%                            4.500000                          5.200000   
50%                            6.400000                          7.200000   
75%                           10.150000                          9.100000   
max                           34.100000                         22.800000   

       Others - 2004-05  Others - 2007-08  All Social Groups - 2004-05  \
count         29.000000         29.000000                    36.000000   
mean           6.744828          6.855172                     8.219444   
std            4.885444          3.857792                     6.816569   
min            0.800000          0.900000                     0.400000   
25%            3.800000          4.500000                     3.700000   
50%            5.300000          5.900000                     6.300000   
75%            7.400000          9.000000                    10.425000   
max           18.400000         18.500000                    27.400000   

       All Social Groups - 2007-08  
count                    36.000000  
mean                      8.136111  
std                       5.350255  
min                       0.600000  
25%                       5.050000  
50%                       6.950000  
75%                       9.050000  
max                      25.200000

In [17]:
data.describe(include=['O'])

Category of States Non Special Category States
count                            36                          36
unique                            4                          36
top     Non Special Category States                       Bihar
freq                             19                           1

#### Checking Missing values in dataFrame for a particluar column

In [18]:
data.isnull()

Category of States  Non Special Category States  \
0                False                        False   
1                False                        False   
2                False                        False   
3                False                        False   
4                False                        False   
5                False                        False   
6                False                        False   
7                False                        False   
8                False                        False   
9                False                        False   
10               False                        False   
11               False                        False   
12               False                        False   
13               False                        False   
14               False                        False   
15               False                        False   
16               False                        False   
17               False                        False   
18               False                        False   
19               False                        False   
20               False                        False   
21               False                        False   
22               False                        False   
23               False                        False   
24               False                        False   
25               False                        False   
26               False                        False   
27               False                        False   
28               False                        False   
29               False                        False   
30               False                        False   
31               False                        False   
32               False                        False   
33               False                        False   
34               False                        False   
35               False                        False   

    Scheduled Castes - 2004-05  Scheduled Castes - 2007-08  \
0                        False                       False   
1                        False                       False   
2                        False                       False   
3                        False                       False   
4                         True                        True   
5                        False                       False   
6                        False                       False   
7                        False                       False   
8                        False                       False   
9                        False                       False   
10                       False                       False   
11                       False                       False   
12                       False                       False   
13                       False                       False   
14                       False                       False   
15                       False                       False   
16                       False                       False   
17                       False                       False   
18                       False                       False   
19                        True                        True   
20                       False                       False   
21                       False                       False   
22                        True                        True   
23                        True                        True   
24                        True                        True   
25                        True                        True   
26                       False                       False   
27                       False                       False   
28                       False                       False   
29                        True                        True   
30                       False                

In [19]:
## count how much NAN value in each column
data.isnull().sum()

Category of States                   0
Non Special Category States          0
Scheduled Castes - 2004-05          10
Scheduled Castes - 2007-08          10
Scheduled Tribes - 2004-05           9
Scheduled Tribes - 2007-08           9
Other Backward Classes - 2004-05     9
Other Backward Classes - 2007-08     9
Others - 2004-05                     7
Others - 2007-08                     7
All Social Groups - 2004-05          0
All Social Groups - 2007-08          0
dtype: int64

#### .iloc() function

.iloc() function enables us to select a particular cell of the dataset that is help us select a value that belongs to a particular row or column from a set of values of a data frame or dataset.

In [20]:
data.iloc[0:10] ## retrieve starting 10 rows

Category of States Non Special Category States  \
0  Non Special Category States              Andhra Pradesh   
1  Non Special Category States                       Assam   
2  Non Special Category States                       Bihar   
3  Non Special Category States                Chhattisgarh   
4  Non Special Category States                       Delhi   
5  Non Special Category States                         Goa   
6  Non Special Category States                     Gujarat   
7  Non Special Category States                     Haryana   
8  Non Special Category States                   Jharkhand   
9  Non Special Category States                   Karnataka   

   Scheduled Castes - 2004-05  Scheduled Castes - 2007-08  \
0                        17.1                        12.8   
1                         9.4                         9.5   
2                        10.6                        10.0   
3                        12.5                         4.3   
4                         NaN                         NaN   
5                        19.2                         0.3   
6                         5.4                         3.6   
7                        10.4                        12.2   
8                         5.9                        16.7   
9                         9.8                         8.5   

   Scheduled Tribes - 2004-05  Scheduled Tribes - 2007-08  \
0                        12.0                         8.2   
1                         4.9                         7.7   
2                        11.2                         7.7   
3                         4.2                         5.8   
4                         NaN                         NaN   
5                         NaN                         NaN   
6                         4.4                         3.5   
7                         NaN                         NaN   
8                         6.9                        10.5   
9                         4.7                         4.9   

   Other Backward Classes - 2004-05  Other Backward Classes - 2007-08  \
0                              10.1                               9.7   
1                               4.3                               6.9   
2                               5.2                               7.0   
3                              10.2                               8.2   
4                               1.5                               4.8   
5                              34.1                               4.6   
6                               4.7                               3.4   
7                               5.9                               9.2   
8                               6.4                               8.9   
9                               6.4                               6.1   

   Others - 2004-05  Others - 2007-08  All Social Groups - 2004-05  \
0               7.4               5.6                         11.2   
1               7.3               6.4                          6.4   
2               5.9               5.3                          6.8   
3               8.1               9.0                          8.2   
4               2.6               2.9                          1.9   
5              17.1               5.9                         19.2   
6               3.4               3.0                          4.5   
7               3.9               2.6                          6.2   
8               7.2              10.9                          6.6   
9               5.3               3.9                          6.7   

   All Social Groups - 2007-08  
0                          9.5  
1                          7.0  
2                          7.4  
3                          6.7  
4                          3.8  
5                          4.9  
6                          3.4  
7                          6.9  
8                         11.2  
9                          6.0

In [21]:
## retrieve first 5 rows and 5th, 6th, 7th columns of data frame
data.iloc[0:5,5:8]

Scheduled Tribes - 2007-08  Other Backward Classes - 2004-05  \
0                         8.2                              10.1   
1                         7.7                               4.3   
2                         7.7                               5.2   
3                         5.8                              10.2   
4                         NaN                               1.5   

   Other Backward Classes - 2007-08  
0                               9.7  
1                               6.9  
2                               7.0  
3                               8.2  
4                               4.8

In [22]:
## retrieve first two columns of data frame with all rows
data.iloc[:,0:2]

Category of States Non Special Category States
0   Non Special Category States              Andhra Pradesh
1   Non Special Category States                       Assam
2   Non Special Category States                       Bihar
3   Non Special Category States                Chhattisgarh
4   Non Special Category States                       Delhi
5   Non Special Category States                         Goa
6   Non Special Category States                     Gujarat
7   Non Special Category States                     Haryana
8   Non Special Category States                   Jharkhand
9   Non Special Category States                   Karnataka
10  Non Special Category States                      Kerala
11  Non Special Category States              Madhya Pradesh
12  Non Special Category States                 Maharashtra
13  Non Special Category States                      Orissa
14  Non Special Category States                      Punjab
15  Non Special Category States                   Rajasthan
16  Non Special Category States                  Tamil Nadu
17  Non Special Category States               Uttar Pradesh
18  Non Special Category States                 West Bengal
19      Special Category States           Arunachal Pradesh
20      Special Category States            Himachal Pradesh
21      Special Category States             Jammu & Kashmir
22      Special Category States                     Manipur
23      Special Category States                   Meghalaya
24      Special Category States                     Mizoram
25      Special Category States                    Nagaland
26      Special Category States                      Sikkim
27      Special Category States                     Tripura
28      Special Category States                 Uttarakhand
29            Union Territories           Andaman & Nicobar
30            Union Territories                  Chandigarh
31            Union Territories        Dadra & Nagar Haveli
32            Union Territories                 Daman & Diu
33            Union Territories                 Lakshadweep
34            Union Territories                  Puducherry
35                    All India                   All India

*** .iloc returns a Pandas Series when one row is selected, and a Pandas DataFrame when multiple rows are selected, or if any column in full is selected.***

#### .loc() function

Pandas loc indexer can be used with DataFrames for two different use cases :-

    a) Selecting rows by label/index
    
    b) Selecting rows with a boolean / conditional lookup

The loc indexer is used with the same syntax as iloc: data.loc[<row selection>, <column selection>] 

In [23]:
## retreive first row data
data.loc[0]

Category of States                  Non Special Category States
Non Special Category States                      Andhra Pradesh
Scheduled Castes - 2004-05                                 17.1
Scheduled Castes - 2007-08                                 12.8
Scheduled Tribes - 2004-05                                   12
Scheduled Tribes - 2007-08                                  8.2
Other Backward Classes - 2004-05                           10.1
Other Backward Classes - 2007-08                            9.7
Others - 2004-05                                            7.4
Others - 2007-08                                            5.6
All Social Groups - 2004-05                                11.2
All Social Groups - 2007-08                                 9.5
Name: 0, dtype: object

In [24]:
## Retrive first 5 rows and 'Scheduled Castes - 2007-08','Scheduled Castes - 2004-05' column from datset
data.loc[0:5,['Scheduled Castes - 2007-08','Scheduled Castes - 2004-05']]

Scheduled Castes - 2007-08  Scheduled Castes - 2004-05
0                        12.8                        17.1
1                         9.5                         9.4
2                        10.0                        10.6
3                         4.3                        12.5
4                         NaN                         NaN
5                         0.3                        19.2

In [25]:
## Retrieve data from datset whose "Category of States" is equal to "Union Territories". 
data.loc[data["Category of States"] == "Union Territories"]

Category of States Non Special Category States  Scheduled Castes - 2004-05  \
29  Union Territories           Andaman & Nicobar                         NaN   
30  Union Territories                  Chandigarh                        13.4   
31  Union Territories        Dadra & Nagar Haveli                         NaN   
32  Union Territories                 Daman & Diu                         NaN   
33  Union Territories                 Lakshadweep                         NaN   
34  Union Territories                  Puducherry                        36.0   

    Scheduled Castes - 2007-08  Scheduled Tribes - 2004-05  \
29                         NaN                         NaN   
30                         4.2                         NaN   
31                         NaN                         7.1   
32                         NaN                         NaN   
33                         NaN                        25.1   
34                        28.7                         NaN   

    Scheduled Tribes - 2007-08  Other Backward Classes - 2004-05  \
29                         NaN                               NaN   
30                         NaN                               NaN   
31                         8.4                               NaN   
32                         NaN                               NaN   
33                        17.0                               NaN   
34                         NaN                              22.2   

    Other Backward Classes - 2007-08  Others - 2004-05  Others - 2007-08  \
29                               NaN              13.4              11.0   
30                               NaN               NaN               NaN   
31                               NaN               NaN               NaN   
32                               NaN               NaN               NaN   
33                               NaN               NaN               NaN   
34                              22.8               NaN               NaN   

    All Social Groups - 2004-05  All Social Groups - 2007-08  
29                         13.4                         10.7  
30                          3.7                          5.9  
31                          6.0                          6.2  
32                          1.6                          1.7  
33                         22.5                         16.7  
34                         27.4                         25.2

In [26]:
## Retrieve data from datset whose "Scheduled Castes - 2004-05" is greater than 10. 
data.loc[data["Scheduled Castes - 2004-05"] > 10]

Category of States Non Special Category States  \
0   Non Special Category States              Andhra Pradesh   
2   Non Special Category States                       Bihar   
3   Non Special Category States                Chhattisgarh   
5   Non Special Category States                         Goa   
7   Non Special Category States                     Haryana   
10  Non Special Category States                      Kerala   
11  Non Special Category States              Madhya Pradesh   
12  Non Special Category States                 Maharashtra   
13  Non Special Category States                      Orissa   
14  Non Special Category States                      Punjab   
16  Non Special Category States                  Tamil Nadu   
18  Non Special Category States                 West Bengal   
27      Special Category States                     Tripura   
30            Union Territories                  Chandigarh   
34            Union Territories                  Puducherry   
35                    All India                   All India   

    Scheduled Castes - 2004-05  Scheduled Castes - 2007-08  \
0                         17.1                        12.8   
2                         10.6                        10.0   
3                         12.5                         4.3   
5                         19.2                         0.3   
7                         10.4                        12.2   
10                        35.0                        27.2   
11                        12.8                        10.2   
12                        17.6                        10.5   
13                        11.1                         9.8   
14                        14.1                        13.1   
16                        22.9                        25.1   
18                        12.5                        13.5   
27                        15.3                        19.7   
30                        13.4                         4.2   
34                        36.0                        28.7   
35                        12.0                        11.9   

    Scheduled Tribes - 2004-05  Scheduled Tribes - 2007-08  \
0                         12.0                         8.2   
2                         11.2                         7.7   
3                          4.2                         5.8   
5                          NaN                         NaN   
7                          NaN                         NaN   
10                        36.5                        29.0   
11                         3.3                         9.0   
12                        11.2                        12.2   
13                         6.3                         6.4   
14                         NaN                         NaN   
16                        26.4                        26.9   
18                        14.4                        14.3   
27                        16.4                        18.2   
30                         NaN                         NaN   
34                         NaN                         NaN   
35                         6.5                         7.5   

    Other Backward Classes - 2004-05  Other Backward Classes - 2007-08  \
0                               10.1                               9.7   
2                                5.2                               7.0   
3                               10.2                               8.2   
5                               34.1                               4.6   
7                                5.9                               9.2   
10                              26.3                              20.2   
11                               5.2                               7.2   
12                               9.3                               6.9   
13                              10.0                               8.4   
14                              10.4                               9.0   
16               

In [27]:
## select rows with id column between 20 and 30, and just return "Scheduled Castes-2004-05","Scheduled Castes - 2007-08","Scheduled Tribes - 2004-05","Scheduled Tribes - 2007-08" columns
data.loc[20:30,["Scheduled Castes - 2004-05","Scheduled Castes - 2007-08","Scheduled Tribes - 2004-05","Scheduled Tribes - 2007-08"]]

Scheduled Castes - 2004-05  Scheduled Castes - 2007-08  \
20                         8.4                         8.8   
21                         2.9                         5.2   
22                         NaN                         NaN   
23                         NaN                         NaN   
24                         NaN                         NaN   
25                         NaN                         NaN   
26                         8.0                        10.3   
27                        15.3                        19.7   
28                         6.0                        11.0   
29                         NaN                         NaN   
30                        13.4                         4.2   

    Scheduled Tribes - 2004-05  Scheduled Tribes - 2007-08  
20                         3.2                         4.1  
21                         3.3                        10.1  
22                         0.2                         3.4  
23                         0.4                         1.4  
24                         0.6                         0.6  
25                         3.3                         7.1  
26                         2.3                         3.1  
27                        16.4                        18.2  
28                         NaN                         NaN  
29                         NaN                         NaN  
30                         NaN                         NaN

#### Dropping any rows with a DataFrame

In [28]:
import numpy as np

In [29]:
df = pd.DataFrame(np.arange(12).reshape(3, 4),columns=['A', 'B', 'C', 'D'])
df

A  B   C   D
0  0  1   2   3
1  4  5   6   7
2  8  9  10  11

In [30]:
### Drop columns
df.drop(['B', 'C'], axis=1)

A   D
0  0   3
1  4   7
2  8  11

In [31]:
### Drop column 'B'and 'C'
df.drop(columns=['B', 'C'])

A   D
0  0   3
1  4   7
2  8  11

In [32]:
### Drop a row by index
df.drop([0, 1])

A  B   C   D
2  8  9  10  11

In [33]:
df

A  B   C   D
0  0  1   2   3
1  4  5   6   7
2  8  9  10  11

In [34]:
df.drop(columns = ['B'],inplace = True)

In [35]:
df

A   C   D
0  0   2   3
1  4   6   7
2  8  10  11

#### dropna() function for deleting NAN values

using dropna() function we delete NAN values containing rows from the dataframe.

Some useful Parameters
------------------------

axis : {0 or 'index', 1 or 'columns'}, default 0
    
    * 0, or 'index' : Drop rows which contain missing values.
        
    * 1, or 'columns' : Drop columns which contain missing value.

how : {'any', 'all'}, default 'any'
    
    Determine if row or column is removed from DataFrame, when we have at least one NA or all NA.

    * 'any' : If any NA values are present, drop that row or column.
        
    * 'all' : If all values are NA, drop that row or column.

thresh : int, optional
    
    Require that many non-NA values.
    
inplace : bool, default False
    
    If True, do operation inplace and return None.

In [36]:
data.head(10)

Category of States Non Special Category States  \
0  Non Special Category States              Andhra Pradesh   
1  Non Special Category States                       Assam   
2  Non Special Category States                       Bihar   
3  Non Special Category States                Chhattisgarh   
4  Non Special Category States                       Delhi   
5  Non Special Category States                         Goa   
6  Non Special Category States                     Gujarat   
7  Non Special Category States                     Haryana   
8  Non Special Category States                   Jharkhand   
9  Non Special Category States                   Karnataka   

   Scheduled Castes - 2004-05  Scheduled Castes - 2007-08  \
0                        17.1                        12.8   
1                         9.4                         9.5   
2                        10.6                        10.0   
3                        12.5                         4.3   
4                         NaN                         NaN   
5                        19.2                         0.3   
6                         5.4                         3.6   
7                        10.4                        12.2   
8                         5.9                        16.7   
9                         9.8                         8.5   

   Scheduled Tribes - 2004-05  Scheduled Tribes - 2007-08  \
0                        12.0                         8.2   
1                         4.9                         7.7   
2                        11.2                         7.7   
3                         4.2                         5.8   
4                         NaN                         NaN   
5                         NaN                         NaN   
6                         4.4                         3.5   
7                         NaN                         NaN   
8                         6.9                        10.5   
9                         4.7                         4.9   

   Other Backward Classes - 2004-05  Other Backward Classes - 2007-08  \
0                              10.1                               9.7   
1                               4.3                               6.9   
2                               5.2                               7.0   
3                              10.2                               8.2   
4                               1.5                               4.8   
5                              34.1                               4.6   
6                               4.7                               3.4   
7                               5.9                               9.2   
8                               6.4                               8.9   
9                               6.4                               6.1   

   Others - 2004-05  Others - 2007-08  All Social Groups - 2004-05  \
0               7.4               5.6                         11.2   
1               7.3               6.4                          6.4   
2               5.9               5.3                          6.8   
3               8.1               9.0                          8.2   
4               2.6               2.9                          1.9   
5              17.1               5.9                         19.2   
6               3.4               3.0                          4.5   
7               3.9               2.6                          6.2   
8               7.2              10.9                          6.6   
9               5.3               3.9                          6.7   

   All Social Groups - 2007-08  
0                          9.5  
1                          7.0  
2                          7.4  
3                          6.7  
4                          3.8  
5                          4.9  
6                          3.4  
7                          6.9  
8                         11.2  
9                          6.0

In [37]:
data.dropna()

Category of States Non Special Category States  \
0   Non Special Category States              Andhra Pradesh   
1   Non Special Category States                       Assam   
2   Non Special Category States                       Bihar   
3   Non Special Category States                Chhattisgarh   
6   Non Special Category States                     Gujarat   
8   Non Special Category States                   Jharkhand   
9   Non Special Category States                   Karnataka   
10  Non Special Category States                      Kerala   
11  Non Special Category States              Madhya Pradesh   
12  Non Special Category States                 Maharashtra   
13  Non Special Category States                      Orissa   
15  Non Special Category States                   Rajasthan   
16  Non Special Category States                  Tamil Nadu   
17  Non Special Category States               Uttar Pradesh   
18  Non Special Category States                 West Bengal   
20      Special Category States            Himachal Pradesh   
21      Special Category States             Jammu & Kashmir   
26      Special Category States                      Sikkim   
27      Special Category States                     Tripura   
35                    All India                   All India   

    Scheduled Castes - 2004-05  Scheduled Castes - 2007-08  \
0                         17.1                        12.8   
1                          9.4                         9.5   
2                         10.6                        10.0   
3                         12.5                         4.3   
6                          5.4                         3.6   
8                          5.9                        16.7   
9                          9.8                         8.5   
10                        35.0                        27.2   
11                        12.8                        10.2   
12                        17.6                        10.5   
13                        11.1                         9.8   
15                         9.1                         8.7   
16                        22.9                        25.1   
17                         5.3                         8.4   
18                        12.5                        13.5   
20                         8.4                         8.8   
21                         2.9                         5.2   
26                         8.0                        10.3   
27                        15.3                        19.7   
35                        12.0                        11.9   

    Scheduled Tribes - 2004-05  Scheduled Tribes - 2007-08  \
0                         12.0                         8.2   
1                          4.9                         7.7   
2                         11.2                         7.7   
3                          4.2                         5.8   
6                          4.4                         3.5   
8                          6.9                        10.5   
9                          4.7                         4.9   
10                        36.5                        29.0   
11                         3.3                         9.0   
12                        11.2                        12.2   
13                         6.3                         6.4   
15                         2.3                         4.5   
16                        26.4                        26.9   
17                         8.4                        11.1   
18                        14.4                        14.3   
20                         3.2                         4.1   
21                         3.3                        10.1   
26                         2.3                         3.1   
27                        16.4                        18.2   
35                         6.5                         7.5   

    Other Backward Classes - 2004-05  Other Backward Classes - 2007-08  \
0        

In [38]:
data.dropna(axis = 1,how = 'any')

Category of States Non Special Category States  \
0   Non Special Category States              Andhra Pradesh   
1   Non Special Category States                       Assam   
2   Non Special Category States                       Bihar   
3   Non Special Category States                Chhattisgarh   
4   Non Special Category States                       Delhi   
5   Non Special Category States                         Goa   
6   Non Special Category States                     Gujarat   
7   Non Special Category States                     Haryana   
8   Non Special Category States                   Jharkhand   
9   Non Special Category States                   Karnataka   
10  Non Special Category States                      Kerala   
11  Non Special Category States              Madhya Pradesh   
12  Non Special Category States                 Maharashtra   
13  Non Special Category States                      Orissa   
14  Non Special Category States                      Punjab   
15  Non Special Category States                   Rajasthan   
16  Non Special Category States                  Tamil Nadu   
17  Non Special Category States               Uttar Pradesh   
18  Non Special Category States                 West Bengal   
19      Special Category States           Arunachal Pradesh   
20      Special Category States            Himachal Pradesh   
21      Special Category States             Jammu & Kashmir   
22      Special Category States                     Manipur   
23      Special Category States                   Meghalaya   
24      Special Category States                     Mizoram   
25      Special Category States                    Nagaland   
26      Special Category States                      Sikkim   
27      Special Category States                     Tripura   
28      Special Category States                 Uttarakhand   
29            Union Territories           Andaman & Nicobar   
30            Union Territories                  Chandigarh   
31            Union Territories        Dadra & Nagar Haveli   
32            Union Territories                 Daman & Diu   
33            Union Territories                 Lakshadweep   
34            Union Territories                  Puducherry   
35                    All India                   All India   

    All Social Groups - 2004-05  All Social Groups - 2007-08  
0                          11.2                          9.5  
1                           6.4                          7.0  
2                           6.8                          7.4  
3                           8.2                          6.7  
4                           1.9                          3.8  
5                          19.2                          4.9  
6                           4.5                          3.4  
7                           6.2                          6.9  
8                           6.6                         11.2  
9                           6.7                          6.0  
10                         25.6                         19.3  
11                          5.6                          8.0  
12                          8.4                          7.3  
13                         10.2                          8.3  
14                          9.7                          8.2  
15                          4.4                          5.1  
16                         15.1                         17.3  
17                          3.7                          5.9  
18                         11.1                         11.6  
19                          1.4                          2.8  
20                          5.7                          6.8  
21                          5.6                          5.9  
22                          1.7                          4.2  
23                          0.4                          1.5  
24                          0.6                          0.6  
25                          3.3            

In [39]:
data.dropna(axis = 1,thresh = 10)

Category of States Non Special Category States  \
0   Non Special Category States              Andhra Pradesh   
1   Non Special Category States                       Assam   
2   Non Special Category States                       Bihar   
3   Non Special Category States                Chhattisgarh   
4   Non Special Category States                       Delhi   
5   Non Special Category States                         Goa   
6   Non Special Category States                     Gujarat   
7   Non Special Category States                     Haryana   
8   Non Special Category States                   Jharkhand   
9   Non Special Category States                   Karnataka   
10  Non Special Category States                      Kerala   
11  Non Special Category States              Madhya Pradesh   
12  Non Special Category States                 Maharashtra   
13  Non Special Category States                      Orissa   
14  Non Special Category States                      Punjab   
15  Non Special Category States                   Rajasthan   
16  Non Special Category States                  Tamil Nadu   
17  Non Special Category States               Uttar Pradesh   
18  Non Special Category States                 West Bengal   
19      Special Category States           Arunachal Pradesh   
20      Special Category States            Himachal Pradesh   
21      Special Category States             Jammu & Kashmir   
22      Special Category States                     Manipur   
23      Special Category States                   Meghalaya   
24      Special Category States                     Mizoram   
25      Special Category States                    Nagaland   
26      Special Category States                      Sikkim   
27      Special Category States                     Tripura   
28      Special Category States                 Uttarakhand   
29            Union Territories           Andaman & Nicobar   
30            Union Territories                  Chandigarh   
31            Union Territories        Dadra & Nagar Haveli   
32            Union Territories                 Daman & Diu   
33            Union Territories                 Lakshadweep   
34            Union Territories                  Puducherry   
35                    All India                   All India   

    Scheduled Castes - 2004-05  Scheduled Castes - 2007-08  \
0                         17.1                        12.8   
1                          9.4                         9.5   
2                         10.6                        10.0   
3                         12.5                         4.3   
4                          NaN                         NaN   
5                         19.2                         0.3   
6                          5.4                         3.6   
7                         10.4                        12.2   
8                          5.9                        16.7   
9                          9.8                         8.5   
10                        35.0                        27.2   
11                        12.8                        10.2   
12                        17.6                        10.5   
13                        11.1                         9.8   
14                        14.1                        13.1   
15                         9.1                         8.7   
16                        22.9                        25.1   
17                         5.3                         8.4   
18                        12.5                        13.5   
19                         NaN                         NaN   
20                         8.4                         8.8   
21                         2.9                         5.2   
22                         NaN                         NaN   
23                         NaN                         NaN   
24                         NaN                         NaN   
25                         NaN                         NaN   
26     

In [40]:
data.dropna(thresh = 10)

Category of States Non Special Category States  \
0   Non Special Category States              Andhra Pradesh   
1   Non Special Category States                       Assam   
2   Non Special Category States                       Bihar   
3   Non Special Category States                Chhattisgarh   
5   Non Special Category States                         Goa   
6   Non Special Category States                     Gujarat   
7   Non Special Category States                     Haryana   
8   Non Special Category States                   Jharkhand   
9   Non Special Category States                   Karnataka   
10  Non Special Category States                      Kerala   
11  Non Special Category States              Madhya Pradesh   
12  Non Special Category States                 Maharashtra   
13  Non Special Category States                      Orissa   
14  Non Special Category States                      Punjab   
15  Non Special Category States                   Rajasthan   
16  Non Special Category States                  Tamil Nadu   
17  Non Special Category States               Uttar Pradesh   
18  Non Special Category States                 West Bengal   
20      Special Category States            Himachal Pradesh   
21      Special Category States             Jammu & Kashmir   
22      Special Category States                     Manipur   
26      Special Category States                      Sikkim   
27      Special Category States                     Tripura   
28      Special Category States                 Uttarakhand   
35                    All India                   All India   

    Scheduled Castes - 2004-05  Scheduled Castes - 2007-08  \
0                         17.1                        12.8   
1                          9.4                         9.5   
2                         10.6                        10.0   
3                         12.5                         4.3   
5                         19.2                         0.3   
6                          5.4                         3.6   
7                         10.4                        12.2   
8                          5.9                        16.7   
9                          9.8                         8.5   
10                        35.0                        27.2   
11                        12.8                        10.2   
12                        17.6                        10.5   
13                        11.1                         9.8   
14                        14.1                        13.1   
15                         9.1                         8.7   
16                        22.9                        25.1   
17                         5.3                         8.4   
18                        12.5                        13.5   
20                         8.4                         8.8   
21                         2.9                         5.2   
22                         NaN                         NaN   
26                         8.0                        10.3   
27                        15.3                        19.7   
28                         6.0                        11.0   
35                        12.0                        11.9   

    Scheduled Tribes - 2004-05  Scheduled Tribes - 2007-08  \
0                         12.0                         8.2   
1                          4.9                         7.7   
2                         11.2                         7.7   
3                          4.2                         5.8   
5                          NaN                         NaN   
6                          4.4                         3.5   
7                          NaN                         NaN   
8                          6.9                        10.5   
9                          4.7                         4.9   
10                        36.5                        29.0   
11                         3.3                         9.0   
12               

#### Replacing a NAN value from a specific value

In [41]:
data.fillna(value = 'Vikas')

Category of States Non Special Category States  \
0   Non Special Category States              Andhra Pradesh   
1   Non Special Category States                       Assam   
2   Non Special Category States                       Bihar   
3   Non Special Category States                Chhattisgarh   
4   Non Special Category States                       Delhi   
5   Non Special Category States                         Goa   
6   Non Special Category States                     Gujarat   
7   Non Special Category States                     Haryana   
8   Non Special Category States                   Jharkhand   
9   Non Special Category States                   Karnataka   
10  Non Special Category States                      Kerala   
11  Non Special Category States              Madhya Pradesh   
12  Non Special Category States                 Maharashtra   
13  Non Special Category States                      Orissa   
14  Non Special Category States                      Punjab   
15  Non Special Category States                   Rajasthan   
16  Non Special Category States                  Tamil Nadu   
17  Non Special Category States               Uttar Pradesh   
18  Non Special Category States                 West Bengal   
19      Special Category States           Arunachal Pradesh   
20      Special Category States            Himachal Pradesh   
21      Special Category States             Jammu & Kashmir   
22      Special Category States                     Manipur   
23      Special Category States                   Meghalaya   
24      Special Category States                     Mizoram   
25      Special Category States                    Nagaland   
26      Special Category States                      Sikkim   
27      Special Category States                     Tripura   
28      Special Category States                 Uttarakhand   
29            Union Territories           Andaman & Nicobar   
30            Union Territories                  Chandigarh   
31            Union Territories        Dadra & Nagar Haveli   
32            Union Territories                 Daman & Diu   
33            Union Territories                 Lakshadweep   
34            Union Territories                  Puducherry   
35                    All India                   All India   

   Scheduled Castes - 2004-05 Scheduled Castes - 2007-08  \
0                        17.1                       12.8   
1                         9.4                        9.5   
2                        10.6                         10   
3                        12.5                        4.3   
4                       Vikas                      Vikas   
5                        19.2                        0.3   
6                         5.4                        3.6   
7                        10.4                       12.2   
8                         5.9                       16.7   
9                         9.8                        8.5   
10                         35                       27.2   
11                       12.8                       10.2   
12                       17.6                       10.5   
13                       11.1                        9.8   
14                       14.1                       13.1   
15                        9.1                        8.7   
16                       22.9                       25.1   
17                        5.3                        8.4   
18                       12.5                       13.5   
19                      Vikas                      Vikas   
20                        8.4                        8.8   
21                        2.9                        5.2   
22                      Vikas                      Vikas   
23                      Vikas                      Vikas   
24                      Vikas                      Vikas   
25                      Vikas                      Vikas   
26                          8                       10.3   
2

#### Replacing 'NAN' value with 'Mean Imputation Method'

In [42]:
def Impute_mean(data,feature,mean):
    data[feature+"_mean"] = data[feature].fillna(mean)

In [43]:
data.head()

Category of States Non Special Category States  \
0  Non Special Category States              Andhra Pradesh   
1  Non Special Category States                       Assam   
2  Non Special Category States                       Bihar   
3  Non Special Category States                Chhattisgarh   
4  Non Special Category States                       Delhi   

   Scheduled Castes - 2004-05  Scheduled Castes - 2007-08  \
0                        17.1                        12.8   
1                         9.4                         9.5   
2                        10.6                        10.0   
3                        12.5                         4.3   
4                         NaN                         NaN   

   Scheduled Tribes - 2004-05  Scheduled Tribes - 2007-08  \
0                        12.0                         8.2   
1                         4.9                         7.7   
2                        11.2                         7.7   
3                         4.2                         5.8   
4                         NaN                         NaN   

   Other Backward Classes - 2004-05  Other Backward Classes - 2007-08  \
0                              10.1                               9.7   
1                               4.3                               6.9   
2                               5.2                               7.0   
3                              10.2                               8.2   
4                               1.5                               4.8   

   Others - 2004-05  Others - 2007-08  All Social Groups - 2004-05  \
0               7.4               5.6                         11.2   
1               7.3               6.4                          6.4   
2               5.9               5.3                          6.8   
3               8.1               9.0                          8.2   
4               2.6               2.9                          1.9   

   All Social Groups - 2007-08  
0                          9.5  
1                          7.0  
2                          7.4  
3                          6.7  
4                          3.8

In [46]:
mean = round(data["Scheduled Castes - 2004-05"].mean(),2)
mean

13.18

In [47]:
Impute_mean(data,"Scheduled Castes - 2004-05",mean)
data.head()

Category of States Non Special Category States  \
0  Non Special Category States              Andhra Pradesh   
1  Non Special Category States                       Assam   
2  Non Special Category States                       Bihar   
3  Non Special Category States                Chhattisgarh   
4  Non Special Category States                       Delhi   

   Scheduled Castes - 2004-05  Scheduled Castes - 2007-08  \
0                        17.1                        12.8   
1                         9.4                         9.5   
2                        10.6                        10.0   
3                        12.5                         4.3   
4                         NaN                         NaN   

   Scheduled Tribes - 2004-05  Scheduled Tribes - 2007-08  \
0                        12.0                         8.2   
1                         4.9                         7.7   
2                        11.2                         7.7   
3                         4.2                         5.8   
4                         NaN                         NaN   

   Other Backward Classes - 2004-05  Other Backward Classes - 2007-08  \
0                              10.1                               9.7   
1                               4.3                               6.9   
2                               5.2                               7.0   
3                              10.2                               8.2   
4                               1.5                               4.8   

   Others - 2004-05  Others - 2007-08  All Social Groups - 2004-05  \
0               7.4               5.6                         11.2   
1               7.3               6.4                          6.4   
2               5.9               5.3                          6.8   
3               8.1               9.0                          8.2   
4               2.6               2.9                          1.9   

   All Social Groups - 2007-08  Scheduled Castes - 2004-05_mean  
0                          9.5                            17.10  
1                          7.0                             9.40  
2                          7.4                            10.60  
3                          6.7                            12.50  
4                          3.8                            13.18

Same way for median and mode Imputation method.

Simply replace mean value with median and mode value of respective column value.

### Group by Method

Any group by operation involves one of the following operation on the original obect.They are :-
    
    1) splitting the object
    
    2) appling a function
    
    3) combining the result

In [48]:
data.groupby("Category of States")

In [49]:
data.groupby("Category of States").count()

Non Special Category States  \
Category of States                                         
All India                                              1   
Non Special Category States                           19   
Special Category States                               10   
Union Territories                                      6   

                             Scheduled Castes - 2004-05  \
Category of States                                        
All India                                             1   
Non Special Category States                          18   
Special Category States                               5   
Union Territories                                     2   

                             Scheduled Castes - 2007-08  \
Category of States                                        
All India                                             1   
Non Special Category States                          18   
Special Category States                               5   
Union Territories                                     2   

                             Scheduled Tribes - 2004-05  \
Category of States                                        
All India                                             1   
Non Special Category States                          15   
Special Category States                               9   
Union Territories                                     2   

                             Scheduled Tribes - 2007-08  \
Category of States                                        
All India                                             1   
Non Special Category States                          15   
Special Category States                               9   
Union Territories                                     2   

                             Other Backward Classes - 2004-05  \
Category of States                                              
All India                                                   1   
Non Special Category States                                19   
Special Category States                                     6   
Union Territories                                           1   

                             Other Backward Classes - 2007-08  \
Category of States                                              
All India                                                   1   
Non Special Category States                                19   
Special Category States                                     6   
Union Territories                                           1   

                             Others - 2004-05  Others - 2007-08  \
Category of States                                                
All India                                   1                 1   
Non Special Category States                19                19   
Special Category States                     8                 8   
Union Territories                           1                 1   

                             All Social Groups - 2004-05  \
Category of States                                         
All India                                              1   
Non Special Category States                           19   
Special Category States                               10   
Union Territories                                      6   

                             All Social Groups - 2007-08  \
Category of States                                         
All India                                              1   
Non Special Category States                           19   
Special Category States                               10   
Union Territories                                      6   

                             Scheduled Castes - 2004-05_mean  
Category of States                                            
All India                                                  1  
Non Special Category States                               19  
Special Category States                                   10  
Union Territories         

In [50]:
data.groupby(["Category of States","Non Special Category States"]).count()

Scheduled Castes - 2004-05  \
Category of States          Non Special Category States                               
All India                   All India                                             1   
Non Special Category States Andhra Pradesh                                        1   
                            Assam                                                 1   
                            Bihar                                                 1   
                            Chhattisgarh                                          1   
                            Delhi                                                 0   
                            Goa                                                   1   
                            Gujarat                                               1   
                            Haryana                                               1   
                            Jharkhand                                             1   
                            Karnataka                                             1   
                            Kerala                                                1   
                            Madhya Pradesh                                        1   
                            Maharashtra                                           1   
                            Orissa                                                1   
                            Punjab                                                1   
                            Rajasthan                                             1   
                            Tamil Nadu                                            1   
                            Uttar Pradesh                                         1   
                            West Bengal                                           1   
Special Category States     Arunachal Pradesh                                     0   
                            Himachal Pradesh                                      1   
                            Jammu & Kashmir                                       1   
                            Manipur                                               0   
                            Meghalaya                                             0   
                            Mizoram                                               0   
                            Nagaland                                              0   
                            Sikkim                                                1   
                            Tripura                                               1   
                            Uttarakhand                                           1   
Union Territories           Andaman & Nicobar                                     0   
                            Chandigarh                                            1   
                            Dadra & Nagar Haveli                                  0   
                            Daman & Diu                                           0   
                            Lakshadweep                                           0   
                            Puducherry                                            1   

                                                         Scheduled Castes - 2007-08  \
Category of States          Non Special Category States                               
All India                   All India                                             1   
Non Special Category States Andhra Pradesh                                        1   
                            Assam                                                 1   
                            Bihar                                                 1   
                            Chhattisgarh                                          1   
                            Delhi                                                 0   
                            Goa                       

### Pandas Concatenation function

Concatenate pandas objects along a particular axis with optional set logic
along the other axes.

Can also add a layer of hierarchical indexing on the concatenation axis, which may be useful if the labels are the same (or overlapping) on the passed axis number.

In [51]:
df1 = pd.DataFrame([['a', 1], ['b', 2]],columns=['letter', 'number'])

In [52]:
df1

letter  number
0      a       1
1      b       2

In [53]:
df2 = pd.DataFrame([['c', 3], ['d', 4]],columns=['letter', 'number'])
df2

letter  number
0      c       3
1      d       4

In [54]:
pd.concat([df1, df2])

letter  number
0      a       1
1      b       2
0      c       3
1      d       4

In [55]:
pd.concat([df1,df2],axis = 1)

letter  number letter  number
0      a       1      c       3
1      b       2      d       4

In [56]:
### Combine ``DataFrame`` objects with overlapping columns and return everything. Columns outside the intersection will be filled with ``NaN`` values.
df3 = pd.DataFrame([['c', 3, 'cat'], ['d', 4, 'dog']],columns=['letter', 'number', 'animal'])
df3

letter  number animal
0      c       3    cat
1      d       4    dog

In [57]:
pd.concat([df1, df3])

letter  number animal
0      a       1    NaN
1      b       2    NaN
0      c       3    cat
1      d       4    dog

In [58]:
### Combine ``DataFrame`` objects with overlapping columns and return only those that are shared by passing ``inner`` to the ``join`` keyword argument.
pd.concat([df1, df3], join="inner")

letter  number
0      a       1
1      b       2
0      c       3
1      d       4

In [59]:
pd.concat([df1, df3], join="outer")

letter  number animal
0      a       1    NaN
1      b       2    NaN
0      c       3    cat
1      d       4    dog

In [60]:
pd.concat([df1, df3], join="outer",ignore_index = True)

letter  number animal
0      a       1    NaN
1      b       2    NaN
2      c       3    cat
3      d       4    dog